node/src/reactor/main_reactor/control.rs

```diff
@@ -137,10 +175,66 @@ impl MainReactor {
+                    info!("Validate: switch to ShutdownForUpgrade");
+                    self.state = ReactorState::ShutdownForUpgrade;
+                    (Duration::ZERO, Effects::new())
+                }
             },
-            ReactorState::Upgrade => {
-                // noop ... patiently wait for the runner to shut down cleanly
-                (DELAY_BEFORE_SHUTDOWN, Effects::new())
+            ReactorState::ShutdownForUpgrade => {
+                let highest_switch_block_era = match self.recent_switch_block_headers.last() {
+                    None => {
+                        return (
+                            Duration::ZERO,
+                            utils::new_shutdown_effect(
+                                "recent_switch_block_headers cannot be empty in Upgrade state",
+                            ),
+                        )
+                    }
+                    Some(block_header) => block_header.era_id(),
+                };
+                match self
+                    .validator_matrix
+                    .validator_weights(highest_switch_block_era)
+                {
+                    Some(validator_weights) => {
+                        let should_stop = match self
+                            .storage
+                            .era_has_sufficient_finality_signatures(&validator_weights)
+                        {
+                            Ok(should_stop) => should_stop,
+                            Err(error) => {
+                                return (
+                                    Duration::ZERO,
+                                    utils::new_shutdown_effect(format!(
+                                        "failed check for sufficient finality signatures: {}",
+                                        error
+                                    )),
+                                )
+                            }
+                        };
+                        let effects = if should_stop {
+                            // Allow a short delay for further finality signatures to arrive and be
+                            // stored.
+                            effect_builder
+                                .set_timeout(DELAY_BEFORE_SHUTDOWN)
+                                .event(|_| {
+                                    MainEvent::ControlAnnouncement(
+                                        ControlAnnouncement::ShutdownForUpgrade,
+                                    )
+                                })
+                        } else {
+                            Effects::new()
+                        };
+                        (DELAY_BEFORE_SHUTDOWN, effects)
+                    }
+                    None => (
+                        Duration::ZERO,
+                        utils::new_shutdown_effect(
+                            "validator_weights cannot be missing in Upgrade state",
+                        ),
+                    ),
+                }
```

Comment:
i'll extract this as part of a general clean up after this PR and my upcoming sync-to-historical PR both merge